lib/luacrypto.c

```diff
@@ -0,0 +1,474 @@
+-- Decrypt:
+local plaintext = c:decrypt(nonce, ciphertext, "myaad")
+print(plaintext)
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/crypto.h>
+#include <crypto/hash.h>
+#include <crypto/aead.h>
+#include <linux/err.h>
+#include <linux/scatterlist.h>
+#include <linux/slab.h>
+#include <linux/hex.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+// Generic hash: hash a string with a given algorithm
+static int luacrypto_hash(lua_State *L)
+{
+	size_t datalen;
+	const char *alg = luaL_checkstring(L, 1);
+	const char *data = luaL_checklstring(L, 2, &datalen);
+
+	struct crypto_shash *tfm;
+	struct shash_desc *desc;
+	int digestsize, ret;
+	u8 *digest;
+
+	tfm = crypto_alloc_shash(alg, 0, 0);
+	if (IS_ERR(tfm))
+		return luaL_error(L, "failed to alloc shash tfm");
+
+	digestsize = crypto_shash_digestsize(tfm);
+	digest = kmalloc(digestsize, GFP_ATOMIC);
+	if (!digest)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc digest buffer");
+	}
+
+	desc = kmalloc(sizeof(*desc) + crypto_shash_descsize(tfm), GFP_ATOMIC);
+	if (!desc)
+	{
+		kfree(digest);
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc desc");
+	}
+	desc->tfm = tfm;
+
+	ret = crypto_shash_digest(desc, data, datalen, digest);
+	crypto_free_shash(tfm);
+	kfree(desc);
+
+	if (ret)
+	{
+		kfree(digest);
+		return luaL_error(L, "digest failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)digest, digestsize);
+	kfree(digest);
+	return 1;
+}
+
+static int luacrypto_hmac(lua_State *L)
+{
+	size_t keylen, datalen;
+	const char *alg = luaL_checkstring(L, 1);
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	const char *data = luaL_checklstring(L, 3, &datalen);
+
+	char hmac_alg[CRYPTO_MAX_ALG_NAME];
+	snprintf(hmac_alg, sizeof(hmac_alg), "hmac(%s)", alg);
+
+	struct crypto_shash *tfm = crypto_alloc_shash(hmac_alg, 0, 0);
+	if (IS_ERR(tfm))
+		return luaL_error(L, "failed to alloc hmac tfm");
+
+	int ret = crypto_shash_setkey(tfm, key, keylen);
+	if (ret)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to set hmac key (%d)", ret);
+	}
+
+	int digestsize = crypto_shash_digestsize(tfm);
+	u8 *digest = kmalloc(digestsize, GFP_ATOMIC);
+	if (!digest)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc digest buffer");
+	}
+
+	struct shash_desc *desc = kmalloc(sizeof(*desc) + crypto_shash_descsize(tfm), GFP_ATOMIC);
+	if (!desc)
+	{
+		kfree(digest);
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc desc");
+	}
+	desc->tfm = tfm;
+
+	ret = crypto_shash_digest(desc, data, datalen, digest);
+	crypto_free_shash(tfm);
+	kfree(desc);
+
+	if (ret)
+	{
+		kfree(digest);
+		return luaL_error(L, "hmac failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)digest, digestsize);
+	kfree(digest);
+	return 1;
+}
+
+static int luacrypto_hex2bin(lua_State *L)
```

Comment:
can't you just use sting.[un]pack? see https://www.lua.org/manual/5.4/manual.html#6.4.2

Reply:
but why binary representation needs to be in ascii? couldn't it be raw memory and then formatting it in ascii hex chars? perhaps I'm missing something..

Reply:
I don’t understand the difference between "raw memory" and "ascii" here: aren’t Lua strings equivalent to what is found in raw memory?

In fact, I mainly use `hex2bin` to work with "test vectors" that are found in RFCs, which use hexadecimal representation.

If I wanted the equivalent of `hex2bin` and `bin2hex` with `string.pack`, I’d need that:
```lua
local function bin2hex(str)
  return string.format(("%x"):rep(#str), string.unpack(("B"):rep(#str), str))
end

local function hex2bin(hex)
  local bytes = { string.unpack(("c2"):rep(#hex / 2), hex) }
  for i = 1, #bytes - 1 do -- string.unpack returns extra value at the end
    bytes[i] = tonumber(bytes[i], 16)
  end
  return string.pack(("B"):rep(#bytes), table.unpack(bytes))
end
```

Reply:
I mean the actual numeric representation in memory.. AFAIK, you could just

```
bin = string.pack("j", 0xFFFFFFFFFFFFFFFF)
hex = string.format("%X", string.unpack("j", bin))
```

Reply:
OK I understand. `hex2bin` would be of interest to parse a command line, but here, I only have to get them as is, not putting them within `" "`.

Reply:
That said, having tried, they’re not equivalent:

```lua
> print(#hex2bin("000102030405060708090a0b0c"))
13
> print(#sp("j", 0x000102030405060708090a0b0c))
8
```

And in fact, replacing `hex2bin("…")` by `sp("j", 0x…)`, tests fail.

Reply:
I think you misunderstood it.. `j` means a `lua_Integer`, each integer has 8 bytes.. so you would need to pass a sequence of `j`s and a list of integers.. or you can do it byte per byte.. my point is that, we don't need to reinvent the wheel in the C binding as Lua already has support for that.. if you would like to define the number as string encoded in ascii for the lua script looking better, you can have your bin2hex/hex2bin defined as you stated above..

Reply:
> my point is that, we don't need to reinvent the wheel in the C binding as Lua already has support for that.. if you would like to define the number as string encoded in ascii for the lua script looking better, you can have your bin2hex/hex2bin defined in Lua as you stated above..

The fun fact here is that, viewing that Linux kernel had hex2bin and bin2hex, I thought: "Well, let’s not reinvent the wheel in Lua"… But I do prefer coding in Lua!

Reply:
Would it be OK to add a `utils.lua` file in `lib/`, to make those functions (and future ones as needed) available from various places?

Reply:
yup.. that's subtle.. sometimes we gonna reuse kernel code, sometimes lua code.. the thing is, to use kernel code in this case you need to traverse lua strings multiple times.. I think one of the main advantages of kernel scripting with Lua is using a full-fledge string lib ;-)